### ISSUE TITLE
Hyperlink exercises and their backmatter hints/answers/solutions in pdf

### ISSUE THREAD

**Issue by oscarlevin** [rbeezer/mathbook#927]
To help students navigate between exercises (or projects) and hints (answers/solutions) in backmatter, it would be nice to have links going both ways.  From hint back to the exercise, this is probably easiest if the number of the hint is a link back to the exercise.  From the exercise to the hint, the exercise number is an option, or a subtle `[hint]` at the very end of the exercise.

If there are multiple places where the hints live, going from exercise to hint might be ill defined, but then the `[hint]` would simpler serve the purpose of letting the reader know that this exercise has a hint.  Another issue is that without the *.backmatter.* switches, I'm not sure how to tell xsltproc to put which links.

Screenshots below.  Live example [here](http://discrete.openmathbooks.org/more/mdm.pdf)

Exercise to hint:
![image](https://user-images.githubusercontent.com/6504596/43324951-969a7688-9172-11e8-9cec-bfdd5a74a2d8.png)

hint to exercise:
![image](https://user-images.githubusercontent.com/6504596/43324997-bbd6e72e-9172-11e8-96dc-b12a73b60ab2.png)

**Comment by rbeezer** [comment]
With `.backmatter` switches, I thought to lock-down **only one instance** of any given hint/answer/solution in print.

I've needed to be very careful about duplicate ids in HTML.  I've been slowly adding some similar infrastructure in LaTeX, but there is a lot left to do.

Maybe back matter solutions need "generated ids" e.g. `target-hint-37` before we can predictably link to them.  Of course, the new flexibility of "solutions" makes duplicates really, really easy to create.

Just off the top of my head, maybe `solutions` needs a flag to say **everything** in this `solutions` is (or is not) a target of such a cross-reference.  Then uniqueness is a source responsibility?  (Seems rare you would really duplicate solutions in print.)

**Comment by oscarlevin** [comment]
I inserted `\hypertarget{a-<exercise-number>}` right after the item number on the hint, and similarly `\hypertarget{q-<exercise-number>}` after the item number on the exercise.  Then hyperlinked to these.  I'm confident you have a better handle on how things are labeled, so there might be a simpler approach.  Either way, it would be the author's responsibility to not have two hints and thus duplicate hypertargets.

The larger concern I have is about whether to put the `[hint]` marker at the end of exercises.  How do we know whether to put `[hint]`, `[sol]`, etc?  If an exercise has a solution in source, it could be that I don't want that solution (or any solutions) to show up in the backmatter.  Is there a way to decide whether solutions appear somewhere by scanning to xml, and set the value of the variable that way?

**Comment by rbeezer** [comment]
> The larger concern I have is about whether to put the [hint] marker at the end of exercises.

Right.  ;-)  I think all the information is there.  Look up the hierarchy for any `solutions` whose scope would include the `exercise`, then examine the three attributes where the author lists hint/answer/solution.  Then also look at any (and all) `backmatter/solutions` similarly.  Similar, but inverted, to the `dry-run` templates to see if headings are necessary.

Man, this stuff gets complicated fast.  But the beauty of XSL is you really can totally explore the whole document.

**Comment by rbeezer** [comment]
For the LaTeX conversion, solutions now link back to the original problem, via their numbers.  Nice suggestion.

Work at 44b06a3e9bcbc6a3bc4e725887d10102a7a8fbe1

Going the other direction seems much harder, though perhaps more useful.  While, in theory, all the necessary information is there, it seems a tall order to follow the necessary path through a `solutions` element to determine if `hint` (say) exists in some collection of solutions.  But maybe there is another "solution"?

Do we want this for HTML?  Should the problem number be a knowl for the problem statement, or a hyperlink back to the actual problem?

**Comment by rbeezer** [comment]
A possible approach:
* Search entire document for hint/answer/solution made visible by a `solutions` element, and bundle up as one gigantic node-set.
* When processing an `exercise` check to see if its hint/answer/solution are in the node-set.  If so drop a pointer by using a reliable naming convention for the reference (augment the `internal-id` scheme).

**Comment by oscarlevin** [comment]
A lower tech solutions would be to have switches: link.solutions, link.answers, link.hints.  The assumption would be that if these were set to "yes" then solutions or answers or hints to everything would be put _somewhere_, (either at the end of divisions or in an appendix) and that would be enough.  It would also be nice to allow a publisher control over whether these links show up anyway.

**Comment by rbeezer** [comment]
We have full control over if hint/answer/solution appear, and where.  Solutions are linking back to original statements.  It is the forward link, from original statement to putative solution, that has required more thought.  A small experiment will determine if the current ideas is feasible enough to ramp up.

Yes, switches controlling these links might be nice.  I guess that would mean another 20.  ;-)

**Comment by oscarlevin** [comment]
Right, the question is how you know to link from the problem to solution at all.  Instead of collecting _all_ hints/answers/solutions, I think it would generally be enough to assume that if any exercise has a solution somewhere, then all exercises (of that type: divisional, checkpoint, project) would, as long as it is in the source.  The decision whether solutions are present could be checked automatically, or set by a switch.  But do we really think that authors would put solutions to exercises in some sections and not in others?

As for HTML, it would definitely be nice to give readers an easy way to get back to the exercise from the solution.  On the other hand, I'm not sure how many books have solutions in the backmatter and not as knowls for the online version.  I'd bet even then, students would just open a second tab for the solutions.

**Comment by rbeezer** [comment]
I see.  I'm not collecting **all** of them, just the ones elected to be visible via a `solutions` element.  That's the tricky bit.

> But do we really think that authors would put solutions to exercises in some sections and not in others?

Yes!

Anyway, I have found that making the code 100% general-purpose (not based on assumptions beyond the schema) leads to more reliable code.  Chunking and numbering are two examples.  Though I am sure one can find exceptions.

I'll see how LaTeX goes before thinking about mimicking it for HTML.

**Comment by rbeezer** [comment]
OK, two-way links implemented at f85921693ca6e64c4c9905d247db48d1a19f1427, plus two or three commits prior.  Key step is identifying **all** the duplicated solutions in advance, at 6b732703438a7e35c7f249f8b7e73a5ac1117b02.

More to do, such as controlling switches, but I think this issue has served its purpose.
